packages/model-viewer/src/features/scene-graph.ts

```diff
@@ -0,0 +1,254 @@
+ * worklet via `modelViewer.worklet` using `postMessage`, much like they would
+ * with any other Web Worker.
+ *
+ * Scene graph worklet scripts must be bestowed capabilities by the author of
+ * the `<model-viewer>` markup. The three capabilities currently available
+ * include:
+ *
+ *  - `messaging`: The ability to communicate with other contexts via
+ *    `postMessage` and `MessageChannel`
+ *  - `fetch`: Access to the global `fetch` method for network operations
+ *  - `material-properties`: The ability to manipulate the basic properties of
+ *    a Material and its associated constructs in the scene graph
+ *
+ * A trivial example of creating a scene graph worklet that can manipulate
+ * material properties looks like this:
+ *
+ * ```html
+ * <model-viewer>
+ *   <script type="experimental-scene-graph-worklet"
+ *       allow="material-properties">
+ *
+ *     console.log('Hello from the scene graph worklet!');
+ *
+ *     self.addEventListener('model-change', () => {
+ *       model.materials[0].pbrMetallicRoughness
+ *         .setBaseColorFactor([1, 0, 0, 1]);
+ *     });
+ *
+ *   </script>
+ * </model-viewer>
+ * ```
+ *
+ * Only one worklet is allowed per `<model-viewer>` at a time. If a new worklet
+ * script is appended to a `<model-viewer>` with a running worklet, a new
+ * worklet will be created and the previous one will be terminated. If there
+ * is more than one worklet script at HTML parse time, the last one in tree
+ * order will be used.
+ *
+ * When a worklet is created, `<model-viewer>` will dispatch a 'worklet-created'
+ * event. At the time that this event is dispatched, the worklet will be created
+ * but the model is not guaranteed to have been made available to the worklet.
+ */
+export const SceneGraphMixin = <T extends Constructor<ModelViewerElementBase>>(
+    ModelViewerElement: T): Constructor<SceneGraphInterface>&T => {
+  class SceneGraphModelViewerElement extends ModelViewerElement {
+    @property({type: Object}) protected[$modelGraft]: ModelGraft|null = null;
+
+    protected[$childListMutationHandler] = (records: Array<MutationRecord>) =>
+        this[$onChildListMutation](records);
+
+    protected[$modelGraftMutationHandler] = (event: Event) =>
+        this[$onModelGraftMutation](event);
+
+    protected[$mutationObserver] =
+        new MutationObserver(this[$childListMutationHandler]);
+
+    protected[$executionContext]: ThreeDOMExecutionContext|null = null;
+
+    /**
+     * A reference to the active worklet if one exists, or else `null`. A
+     * worklet is not created until a scene graph worklet script has been
+     * detected as a child of this `<model-viewer>`.
+     */
+    get worklet() {
+      const executionContext = this[$executionContext];
+      return executionContext != null ? executionContext.worker : null;
+    }
+
+    connectedCallback() {
+      super.connectedCallback();
+
+      this[$mutationObserver].observe(this, {childList: true});
```

Comment:
TL;DR IE supports workers. We will need the ShadyDOM version of `MutationObserver`. At this time, 3DOM itself is not known to be compatible with IE. I would like to do the work to fix it as a follow-on task.

Longer version:

IE11 depends on a code transformation step where we transform ES6+ JavaScript to ES5-equivalent JavaScript. This gets a little tricky in the 3DOM case, because we are relying on the strings produced by `Function.prototype.toString` in order to assemble a viable Worker script.  This is just a hypothetical, but consider the following two chunks of code:

```javascript
function greet(message) {
  console.log('Hey!', message);
}

// ###

var stringified = greet.toString();

/* Value of "stringified":

function greet(message) {
  console.log('Hey!', message);
}
*/
```

```javascript
var greet = (function() {
  var callout = 'Hey!';
  return function(message) {
    console.log(callout, message);
  };
})();

// ###

var stringified = greet.toString();

/* Value of "stringified":

function(message) {
  console.log(callout, message);
};
*/
```

So the bottom line is that we need to make sure that an ES6->ES5 transform does not inadvertently produce closure scopes or similar JavaScript-isms that cannot be properly stringified.